### PR TITLE
PS-6024 Integrate pquery3 with old pquery framework

### PR DESCRIPTION
**Pull request by mohitj1988**

### CODE REVIEW

File: pquery-run.sh [hunk 1]
```diff
@@ -338,7 +339,7 @@ if [[ $PXC -eq 1 ]];then
   rm -rf ${BASEDIR}/my.cnf
   echo "[mysqld]" > ${BASEDIR}/my.cnf
   echo "basedir=${BASEDIR}" >> ${BASEDIR}/my.cnf
-  echo "wsrep-debug=1" >> ${BASEDIR}/my.cnf
```
Comment (rameshvs02) [inline pquery-run.sh]: Please revert this change, wsrep-debug=ON is not supported in PXC-8.0.

File: pquery-run.sh [hunk 2]
```diff
@@ -447,35 +448,14 @@ pxc_startup(){
...
-    echo "ssl-ca = ${WORKDIR}/cert/ca.pem" >> ${DATADIR}/n${i}.cnf
-    echo "ssl-cert = ${WORKDIR}/cert/client-cert.pem" >> ${DATADIR}/n${i}.cnf
-    echo "ssl-key = ${WORKDIR}/cert/client-key.pem" >> ${DATADIR}/n${i}.cnf
-    echo "[sst]" >> ${DATADIR}/n${i}.cnf
-    echo "encrypt = 4" >> ${DATADIR}/n${i}.cnf
-    echo "ssl-ca = ${WORKDIR}/cert/ca.pem" >> ${DATADIR}/n${i}.cnf
-    echo "ssl-cert = ${WORKDIR}/cert/server-cert.pem" >> ${DATADIR}/n${i}.cnf
-    echo "ssl-key = ${WORKDIR}/cert/server-key.pem" >> ${DATADIR}/n${i}.cnf    
+
     if [ "$IS_STARTUP" == "startup" ]; then
```
Comment (rameshvs02) [inline pquery-run.sh]: pxc_encrypt_cluster_traffic is ON by default in PXC-8.0 so we need to add SSL configuration in cnf file. If we are not using encryption we should turn off pxc_encrypt_cluster_traffic option.

File: pquery-run.sh [hunk 3]
```diff
@@ -447,35 +448,14 @@ pxc_startup(){
...
       ${MID} --datadir=$node  > ${WORKDIR}/startup_node1.err 2>&1 || exit 1;
     fi
   done
-  if check_for_version $MYSQL_VERSION "8.0.0" ; then
-    if [ "$IS_STARTUP" == "startup" ]; then
-	  mkdir ${WORKDIR}/cert
-	  cp ${WORKDIR}/node1.template/*.pem ${WORKDIR}/cert/
-    fi
-  fi
```
Comment (rameshvs02) [inline pquery-run.sh]: pxc_encrypt_cluster_traffic is ON by default in PXC-8.0 so we need to add SSL configuration in cnf file. If we are not using encryption we should turn off pxc_encrypt_cluster_traffic option.

File: pquery-run.sh (hunk 1 above)
Comment (mohitj1988) [reply pquery-run.sh]: Accidently these PXC changes crept in. Reverted.

File: pquery-run.sh (hunk 2 above)
Comment (mohitj1988) [reply pquery-run.sh]: Accidently these changes crept in. Reverted.

File: pquery-run.sh (hunk 3 above)
Comment (mohitj1988) [reply pquery-run.sh]: Accidently these changes crept in. Reverted.

File: pquery3-run.conf [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+# in pquery-run.sh to match!                                                   #
+# Please do not change this file unless your name is listed above. As a        #
+# template/default, it is the only configuration file for pquery v3.0 which    #
+# is maintained whenever pquery-run.sh is updated.                             #
+################################################################################
+
+# ################### User configurable variables: generics 1 ##################
+# pquery3-ps: Percona Server, pquery3-ms: MySQL                                #
+################################################################################
+PQUERY_BIN=$HOME/pquery_n/src/pquery3-ps
```
Comment (mohitj1988) [inline pquery3-run.conf]: Please note this is incorrect at the moment. Once we have a new repo for pquery3, we can decide upon the path. Also this path can vary from user to user

File: pquery3-run.conf [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+SAVE_SQL=0
+
+################################################################################
+# Store a copy of the INFILE SQL input file in the work directory (not related #
+# to per-trial SQL)                                                            #
+################################################################################
+STORE_COPY_OF_INFILE=0
+
+################################################################################
+# x sec max trial runtime within which pquery tries to process                 #
```
Comment (rahulmalik87) [inline pquery3-run.conf]: give some comment i.e. with every step runtime will increase.

File: pquery-run.sh [hunk 6]
```diff
@@ -1352,13 +1358,22 @@ pquery_test(){
...
+            echoit "Loading metadata from ${WORKDIR}/step_$((${TRIAL}-1)).dll ..."
+          fi
+        else
+          echoit "Taking ${MULTI_THREADED_TESTC_LINES} lines randomly from ${INFILE} as testcase for this multi-threaded trial..."
+        fi
         shuf --random-source=/dev/urandom ${INFILE} | head -n${MULTI_THREADED_TESTC_LINES} > ${RUNDIR}/${TRIAL}/${TRIAL}.sql
         SQL_FILE="--infile=${RUNDIR}/${TRIAL}/${TRIAL}.sql"
       fi
-      # Debug echo "-------"; cat ${RUNDIR}/${TRIAL}/${TRIAL}.sql; echo "-------"
+      PQUERY_RUN_TIMEOUT=$[ ${PQUERY_RUN_TIMEOUT} + 10 ]
```
Comment (rahulmalik87) [inline pquery-run.sh]: make this 10 as a new option.

File: pquery3-run.conf [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+# Please do not change this file unless your name is listed above. As a        #
+# template/default, it is the only configuration file for pquery v3.0 which    #
+# is maintained whenever pquery-run.sh is updated.                             #
+################################################################################
+
+# ################### User configurable variables: generics 1 ##################
+# pquery3-ps: Percona Server, pquery3-ms: MySQL                                #
+################################################################################
+PQUERY_BIN=$HOME/pquery_n/src/pquery3-ps
+PQUERY_TOOL_NAME=pquery3-ps
```
Comment (rameshvs02) [inline pquery3-run.conf]: Can't we read pquery tool name using `basename $PQUERY_BIN` command?

Comment (mohitj1988) [reply pquery3-run.conf]: Good suggestion. I will do the changes

File: pquery-run.sh (hunk 6 above)
Comment (mohitj1988) [reply pquery-run.sh]: Done

File: pquery3-run.conf (hunk 5 above)
Comment (mohitj1988) [reply pquery3-run.conf]: Added comments in the conf file. Also added console logs about it

**Comment by rameshvs02** [comment]
lgtm

File: pquery-run.sh [hunk 8]
```diff
@@ -1352,13 +1358,21 @@ pquery_test(){
...
+        else
+          echoit "Taking ${MULTI_THREADED_TESTC_LINES} lines randomly from ${INFILE} as testcase for this multi-threaded trial..."
+        fi
         shuf --random-source=/dev/urandom ${INFILE} | head -n${MULTI_THREADED_TESTC_LINES} > ${RUNDIR}/${TRIAL}/${TRIAL}.sql
         SQL_FILE="--infile=${RUNDIR}/${TRIAL}/${TRIAL}.sql"
       fi
-      # Debug echo "-------"; cat ${RUNDIR}/${TRIAL}/${TRIAL}.sql; echo "-------"
       if [[ ${PXC} -eq 0 && ${GRP_RPL} -eq 0 ]]; then
-        ${PQUERY_BIN} ${SQL_FILE} --database=test --threads=${THREADS} --queries-per-thread=${QUERIES_PER_THREAD} --logdir=${RUNDIR}/${TRIAL} --log-all-queries --log-failed-queries --user=root --socket=${RUNDIR}/${TRIAL}/socket.sock >${RUNDIR}/${TRIAL}/pquery.log 2>&1 &
+        echo "${PQUERY_BIN} ${SQL_FILE} --database=test --threads=${THREADS} --queries-per-thread=${QUERIES_PER_THREAD} --logdir=${RUNDIR}/${TRIAL} --log-all-queries --log-failed-queries --user=root --socket=${RUNDIR}/${TRIAL}/socket.sock --seed $RANDOMD --step ${TRIAL} --metadata-path ${WORKDIR}/ --seconds ${PQUERY_RUN_TIMEOUT} ${DYNAMIC_QUERY_PARAMETER}";
```
Comment (rahulmalik87) [inline pquery-run.sh]: do not enable log-all-queries by default. let it pass through DYNAMIC_QUERY_PARAMETER

Comment (rahulmalik87) [reply pquery-run.sh]: if seed option is not provided by the user then use a random seed.

File: pquery-run.sh [hunk 9]
```diff
@@ -1408,6 +1422,9 @@ pquery_test(){
         if [ $X -ge ${PQUERY_RUN_TIMEOUT} ]; then
           echoit "${PQUERY_RUN_TIMEOUT}s timeout reached. Terminating this trial..."
           TIMEOUT_REACHED=1
+          echoit "TIMEOUT_INCREMENT option was enabled and set to ${TIMEOUT_INCREMENT} sec"
+          echoit "${TIMEOUT_INCREMENT}s will be added to the next trial timeout."
```
Comment (rahulmalik87) [inline pquery-run.sh]: can we move this and show in the starting, running the test for N seconds. 
would be more relevant.

Comment (mohitj1988) [reply pquery-run.sh]: It will be seen in the beginning of every trial. Reason why it is put at the end of the iteration is, we want the first trial to run without increment which was happening earlier

File: pquery-run.sh (hunk 8 above)
Comment (mohitj1988) [reply pquery-run.sh]: Done

File: pquery-run.sh [hunk 10]
```diff
@@ -1657,6 +1691,16 @@ pquery_test(){
         echoit "Saving full trial outcome (as SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY=0 and so trials are saved irrespective of whether an issue was detected or not)"
         savetrial
         TRIAL_SAVED=1
+      elif [ ${SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY} -eq 1 ]; then
```
Comment (rahulmalik87) [inline pquery-run.sh]: behaviour with 2.0 should be the same.  will it impact the existing behaviour?

Comment (mohitj1988) [reply pquery-run.sh]: Introduced a new variable which is enabled when pquery3 is used. Behaviour of pquery 2.0 is not changed as part of this integration

File: pquery-run.sh [hunk 11]
```diff
@@ -1352,14 +1372,25 @@ pquery_test(){
...
+          else
+            echoit "Loading metadata from ${WORKDIR}/step_$((${TRIAL}-1)).dll ..."
+          fi
+        else
+          echoit "Taking ${MULTI_THREADED_TESTC_LINES} lines randomly from ${INFILE} as testcase for this multi-threaded trial..."
+        fi
         shuf --random-source=/dev/urandom ${INFILE} | head -n${MULTI_THREADED_TESTC_LINES} > ${RUNDIR}/${TRIAL}/${TRIAL}.sql
         SQL_FILE="--infile=${RUNDIR}/${TRIAL}/${TRIAL}.sql"
       fi
-      # Debug echo "-------"; cat ${RUNDIR}/${TRIAL}/${TRIAL}.sql; echo "-------"
```
Comment (rahulmalik87) [inline pquery-run.sh]: since we are not testing/fixing PXC as part of this JIRA. 
Better to ${PQUERY3} -eq 0 ] inside if [[ ${PXC} -eq 0 && ${GRP_RPL} -eq 0 ]

File: pquery-run.sh [hunk 12]
```diff
@@ -1657,6 +1695,16 @@ pquery_test(){
         echoit "Saving full trial outcome (as SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY=0 and so trials are saved irrespective of whether an issue was detected or not)"
         savetrial
         TRIAL_SAVED=1
+      elif [[ ${SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY} -eq 1 && ${PQUERY3} -eq 1 ]]; then
+        if [ ${TRIAL} -gt 1 ]; then
+          echoit "Saving last trial outcome (as SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY=1)"
```
Comment (rahulmalik87) [inline pquery-run.sh]: remove SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY. only sufficient condition is PQUERY3 and remove 
echoit "Saving last trial outcome (as SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY=1)"

File: pquery-run.sh [hunk 13]
```diff
@@ -1657,6 +1695,16 @@ pquery_test(){
...
+          savetrial
```
Comment (rahulmalik87) [inline pquery-run.sh]: you can move savetrial, TRIAL_SAVED=1 out of if-else.

File: pquery-run.sh (hunk 12 above)
Comment (mohitj1988) [reply pquery-run.sh]: We cannot exclude SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY
If you see the old pquery code, they were checking for this condition
 SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY -eq 0
If yes, then TRIAL_SAVED is set to 1, and by doing this removetrial function is not invoked
There was no check for SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY -eq 1, and the reason was, they had no intention to save anything
But now, we have an additional contraint that when SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY -eq 1, we have to not delete everything, but retain the last trial
Hence we need to have  a check in-place

File: pquery-run.sh (hunk 13 above)
Comment (mohitj1988) [reply pquery-run.sh]: We cannot exclude SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY
If you see the old pquery code, they were checking for this condition
 SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY -eq 0
If yes, then TRIAL_SAVED is set to 1, and by doing this removetrial function is not invoked
There was no check for SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY -eq 1, and the reason was, they had no intention to save anything
But now, we have an additional contraint that when SAVE_TRIALS_WITH_CORE_OR_VALGRIND_ONLY -eq 1, we have to not delete everything, but retain the last trial
Hence we need to have  a check in-place

Comment (rahulmalik87) [reply pquery-run.sh]: an additional constraint is only PQUERY3 if it is set then save the last trial.
Unless we are not saving the all trials.
